Add module files for building SCM with spack-stack on Derecho, Hera, Jet, Orion

mkavulich opened the pull request:
This PR introduces modulefiles for building SCM for Derecho (Intel) and Hera (Intel, GNU). It should be fairly easy to add analogous modulefiles for other EPIC-supported platforms, so let me know if that's desired.

I ran the regression test suite and there were some differences on Hera as expected. These differences were almost entirely at the precision noise level (<1e-10) except for a tests that had isolated significant differences. The vast majority of diffs across all fields and all tests were exactly 0. Differences from the baseline (compiled from top of develop with the old shell environment files) can be found in the files in the following directories if anyone wants to take a closer look:

 - /scratch1/BMC/gmtb/CCPP_regression_testing/NCAR_scm/modulefile_updates/ccpp-scm/test/artifact-release_intel
 - /scratch1/BMC/gmtb/CCPP_regression_testing/NCAR_scm/modulefile_updates/gnu/ccpp-scm/test/artifact-release_gnu

Documentation has been updated in the `.tex` files, but I haven't been able to re-build the PDF yet. For now instructions for building on Derecho are here:

https://docs.google.com/document/d/1Wg5dBIzwhjoYf6BhgmsUJczPEfxD3dA2yTRS5ftSICk/edit

grantfirl commented on scm/etc/modules/hera_intel.lua:
```diff
@@ -0,0 +1,46 @@
+prepend_path("MODULEPATH","/contrib/sutils/modulefiles")
+load("sutils")
+
+prepend_path("MODULEPATH", "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.4.1/envs/unified-env/install/modulefiles/Core")
```
Can we use the same version of spack-stack as in ufs-weather-model: https://github.com/ufs-community/ufs-weather-model/blob/develop/modulefiles/ufs_hera.intel.lua?

climbfuji replied:
Yes - ufs will move to 1.5.1 shortly. Do you want me to create a PR and update to 1.5.1? This way you don't waste your time if something goes less smooth than I was bragging.

grantfirl replied:
@mkavulich Can you update this PR at least to spack-stack-1.5.0 and then once ufs moves to 1.5.1, we can try to follow suit. I'd like to keep the SCM using the same stack as ufs-wm going forward, we just haven't been paying attention for a while due to not having a release.

grantfirl commented on scm/etc/modules/derecho_intel.lua:
```diff
@@ -0,0 +1,47 @@
+load(pathJoin("intel-classic", os.getenv("intel_classic_ver") or "2023.0.0"))
+load(pathJoin("cray-mpich", os.getenv("cray_mpich_ver") or "8.1.25"))
+
+prepend_path("MODULEPATH","/glade/work/epicufsrt/contrib/derecho/hpc-stack/intel-classic-2023.0.0/modulefiles/stack")
```
I think that there is still a reliance on some non-standard python packages, like f90nml. Would we still need to create a python environment on top of the one from spack-stack in this case?

climbfuji replied:
Give me a few minutes to answer your question please!

mkavulich replied:
@climbfuji Thanks for the link; last I had checked spack-stack was still not supported on Derecho so I stuck with the libraries I knew would work. I will see if those work for the SCM build.

mkavulich replied:
@climbfuji I wasn't able to get spack-stack to work on Derecho. I receive the following error when attempting to load the spack-stack modules:

```Lmod has detected the following error:  Cannot load module "stack-cray-mpich/8.1.25" without these module(s) loaded:
   libfabric/1.15.2.0

While processing the following module(s):
    Module fullname          Module Filename
    ---------------          ---------------
    stack-cray-mpich/8.1.25  /glade/work/epicufsrt/contrib/spack-stack/derecho/spack-stack-1.5.1/envs/unified-env/install/modulefiles/intel/2021.10.0/stack-cray-mpich/8.1.25.lua
    derecho_intel            /glade/derecho/scratch/kavulich/SCM/PR_406/ccpp-scm/scm/etc/modules/derecho_intel.lua
```

Which doesn't make sense to me, because `libfabric/1.15.2.0` is supposed to be loaded *inside* the `stack-cray-mpich/8.1.25` module (and trying to manually load it right before that step doesn't work either). I'll admit I'm a little unclear on the finer details of these modules though, is it possible I'm not loading these in the right order or something?  The modulefile I'm using is below:

```
help([[
This module loads libraries for building the CCPP Single-Column Model on
the CISL machine Derecho (Cray) using Intel-classic-2023.0.0
]])

whatis([===[Loads libraries needed for building the CCPP SCM on Derecho ]===])

load(pathJoin("cmake", os.getenv("cmake_ver") or "3.26.3"))
load(pathJoin("ncarenv", os.getenv("ncarenv_ver") or "23.06"))
load(pathJoin("craype", os.getenv("craype_ver") or "2.7.20"))

prepend_path("MODULEPATH","/glade/work/epicufsrt/contrib/spack-stack/derecho/spack-stack-1.5.1/envs/unified-env/install/modulefiles/Core")
load("stack-intel/2021.10.0")
load("stack-cray-mpich/8.1.25")
load("stack-python/3.10.8")

load("bacio/2.4.1")
load("sp/2.3.3")
load("w3emc/2.9.2")

setenv("CC","cc")
setenv("FC","ftn")
setenv("CXX","CC")

setenv("CMAKE_C_COMPILER","cc")
setenv("CMAKE_CXX_COMPILER","CC")
setenv("CMAKE_Fortran_COMPILER","ftn")
setenv("CMAKE_Platform","derecho.intel")
```

climbfuji replied:
Did you use the "preamble" from spack-stack? In particular, I think you need (see https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#ncar-wyoming-derecho):
```
module purge
# ignore that the sticky module ncarenv/... is not unloaded
export LMOD_TMOD_FIND_FIRST=yes
module load ncarenv/23.09
module use /glade/work/epicufsrt/contrib/spack-stack/derecho/modulefiles
```
and then the module use/load statements for either GNU or Intel.

mkavulich replied:
@climbfuji Sorry about that....writing out my last message inspired me to figure out the first issue (you are right I had not done the first step of appending the modulepath for spack stack modules). I am building now and will let you know if these tests succeed.

One thing I noticed is that the `w3emc` libraries are listed without a version in the Derecho spack-stack. This seems odd to me, do you know why this is?

mkavulich replied:
> I think that there is still a reliance on some non-standard python packages, like f90nml. Would we still need to create a python environment on top of the one from spack-stack in this case?

@grantfirl It looks like we do need to keep our own python environment, at least for now (the spack-stack environment does not contain f90nml as you expected).

climbfuji replied:
It does, `module load py-f90nml`

mkavulich replied:
Thanks again @climbfuji. So all the python packages in spack-stack are loaded via modules? I didn't think to look there, it looks like a lot of good packages 👍

climbfuji replied:
Yes, all of them are modules and all python packages start with `py-`.

climbfuji replied:
You can extend these spack python environments in case packages are missing to load whatever you can as modules from spack-stack, then create a virtual environment (`python3 -m venv venv`) and then install missing packages via pip (`python3 -m pip install NAME`). This way, all the spack-stack Python utilities are used unless there are version conflicts.

climbfuji commented:
I'll push a branch with an example spack-stack file for hera/intel shortly. The only thing that's missing I think is the optional doyxgen. We can consider adding that to spack-stack (but I remember it's a bit of a tricky package). We can also consider adding a separate template for the scm so that users who just want that (and not the full ufs-weather-model) only have to build a few libraries. But then again, the ufs-weather-model template should be good enough.

mkavulich commented:
@DomHeinzeller @grantfirl I have updated the new modules to all use spack-stack 1.5.1, and also added one for Derecho GNU. I did re-run the regression tests for Hera Intel/GNU and they all passed but I have not done a comparison with the main branch baseline for Hera to ensure close-ish results; I can do that if you'd like but I just haven't had time yet.

I also rebased my branch on the latest develop to fix the CI tests, all now seem to be passing.

grantfirl commented:
@mkavulich I've tried loading hera_intel on Hera with this code, and it works fine for me. I'm guessing that we'll need to tell folks to manually set the SCM_ROOT variable or does it make any sense to try to set it via the lua file?

grantfirl commented:
@mkavulich @dustinswales How could we use spack-stack for the CI tests? I'm guessing that if we want to switch that over too, we'll do that in a separate PR?

mkavulich commented:
> @mkavulich I've tried loading hera_intel on Hera with this code, and it works fine for me. I'm guessing that we'll need to tell folks to manually set the SCM_ROOT variable or does it make any sense to try to set it via the lua file?

I've been meaning to talk to you about that. It seems to me like using this variable is unnecessary complexity, if this is set automatically through a setup script or modulefile why don't we just set it directly in the python script?

grantfirl commented:
> > @mkavulich I've tried loading hera_intel on Hera with this code, and it works fine for me. I'm guessing that we'll need to tell folks to manually set the SCM_ROOT variable or does it make any sense to try to set it via the lua file?
> 
> I've been meaning to talk to you about that. It seems to me like using this variable is unnecessary complexity, if this is set automatically through a setup script or modulefile why don't we just set it directly in the python script?

Ya, that should work fine. The whole idea of having SCM_ROOT in the first place was to allow for flexibility with respect to where executables are stored and where the output goes. In the run script, we could check if the `SCM_ROOT` environment variable exists. If so, use it, if not, find the top level ccpp-scm directory above where the run script is being called and use that.

climbfuji commented:
> @mkavulich @dustinswales How could we use spack-stack for the CI tests? I'm guessing that if we want to switch that over too, we'll do that in a separate PR?
> 
> For example, see https://github.com/JCSDA/spack-stack/blob/release/1.5.1/.github/workflows/ubuntu-ci-x86_64.yaml for setting up the environment?

You could try to pull the containers we create for JEDI CI, they should have all the dependencies you need (but I agree that making this or any other solution a separate PR is better)

grantfirl commented:
@mkavulich I'm running into issues on Derecho. It apparently can't find NetCDF-fortran. You don't get this error?

CMake Error at /glade/work/grantf/ccpp-scm/CMakeModules/Modules/FindNetCDF.cmake:246 (message):
  Unable to properly find NetCDF.  Found static libraries at:
  /glade/work/grantf/ccpp-scm/scm/src/NetCDF_Fortran_LIBRARY-NOTFOUND but
  could not run nc-config:
Call Stack (most recent call first):
  CMakeLists.txt:67 (find_package)


CMake Error at /glade/u/apps/derecho/23.09/spack/opt/spack/cmake/3.26.3/gcc/7.5.0/k34x/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find NetCDF (missing: Fortran) (found version "4.9.2")
Call Stack (most recent call first):
  /glade/u/apps/derecho/23.09/spack/opt/spack/cmake/3.26.3/gcc/7.5.0/k34x/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /glade/work/grantf/ccpp-scm/CMakeModules/Modules/FindNetCDF.cmake:312 (find_package_handle_standard_args)
  CMakeLists.txt:67 (find_package)

grantfirl commented:
@mkavulich FYI, if I add the netCDF load commands to the Derecho lua files, everything works fine for me.

grantfirl commented on scm/etc/modules/derecho_gnu.lua:
```diff
@@ -0,0 +1,28 @@
+load("stack-python/3.10.8")
+load("py-f90nml")
+load("py-netcdf4/1.5.8")
+
```
I think that we need this here:
`load("netcdf-c/4.9.2")`
`load("netcdf-fortran/4.6.0")`

grantfirl commented on scm/etc/modules/derecho_intel.lua:
```diff
@@ -0,0 +1,32 @@
+load("stack-python/3.10.8")
+load("py-f90nml")
+load("py-netcdf4/1.5.8")
+
```
I think that we need this here:
load("netcdf-c/4.9.2")
load("netcdf-fortran/4.6.0")

climbfuji replied:
Lines 21 and 22

grantfirl reviewed (changes requested):
I'm fine with everything except NetCDF on Derecho (see comments).

mkavulich commented:
@grantfirl Thanks for testing this out, I have had some testing frustrations because `module purge` does not seem to actually fully purge my environment when running different tests (or maybe there's some cmake cacheing going on that I don't understand? I'd say 50/50 a platform error or user error). I fully logged out and logged back in and started with a fresh clone and purged environment, and was able to replicate your issue. I pushed those changes for the Derecho Intel and GNU modulefiles, and also added a default value for SCM_ROOT per our other conversation.

climbfuji commented:
> @grantfirl Thanks for testing this out, I have had some testing frustrations because `module purge` does not seem to actually fully purge my environment when running different tests (or maybe there's some cmake cacheing going on that I don't understand? I'd say 50/50 a platform error or user error). I fully logged out and logged back in and started with a fresh clone and purged environment, and was able to replicate your issue. I pushed those changes for the Derecho Intel and GNU modulefiles, and also added a default value for SCM_ROOT per our other conversation.

To me this is a bit suspicious. The spack `py-netcdf4` does depend on netcdf-c, so if that module doesn't get loaded automatically when `py-netcdf4` is loaded, then something is off. Loading `netcdf-fortran` should also automatically load `netcdf-c` if it isn't loaded yet. But of course it doesn't harm to list the version explicitly.

climbfuji commented on scm/etc/modules/hera_gnu.lua:
```diff
@@ -0,0 +1,29 @@
+load("py-f90nml")
+load("py-netcdf4/1.5.8")
+
+
```
Super picky, but remove extra blank line?

climbfuji commented on scm/etc/modules/hera_intel.lua:
```diff
@@ -0,0 +1,29 @@
+load("py-f90nml")
+load("py-netcdf4/1.5.8")
+
+
```
Ditto

mkavulich commented:
> To me this is a bit suspicious. The spack `py-netcdf4` does depend on netcdf-c, so if that module doesn't get loaded automatically when `py-netcdf4` is loaded, then something is off. Loading `netcdf-fortran` should also automatically load `netcdf-c` if it isn't loaded yet. But of course it doesn't harm to list the version explicitly.

I agree that it is suspicious that this issue is occurring. There appears to be something going on with different hdf5 versions compared to the system default. When you don't run a `module purge` prior to running, this is the result: of loading the current `derecho_gnu.lua` module:

```
> module load derecho_gnu

Lmod is automatically replacing "intel/2023.0.0" with "gcc/12.2.0".

Lmod Warning: 
------------------------------------------------------------------------------------------------------
The following dependent module(s) are not currently loaded: hdf5/1.14.0 (required by:
py-netcdf4/1.5.8, netcdf-c/4.9.2)
------------------------------------------------------------------------------------------------------




Due to MODULEPATH changes, the following have been reloaded:
  1) cray-mpich/8.1.25     2) craype/2.7.20     3) hdf5/1.12.2     4) ncarcompilers/1.0.0     5) netcdf/4.9.2

The following have been reloaded with a version change:
  1) ncarenv/23.06 => ncarenv/23.09
```

Running `module purge` prior to loading makes the load go much more smoothly:

```
> module load derecho_intel

The following have been reloaded with a version change:
  1) ncarenv/23.06 => ncarenv/23.09

```

Now, both of those do work, but it maybe the warning does give some hint as to why those netcdf modules need to be explicitly loaded.

climbfuji commented:
You have to follow exactly the steps in https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#ncar-wyoming-derecho unless you want to set yourself up for trouble:
```
module purge
# ignore that the sticky module ncarenv/... is not unloaded
export LMOD_TMOD_FIND_FIRST=yes
module load ncarenv/23.09
module use /glade/work/epicufsrt/contrib/spack-stack/derecho/modulefiles
module load ecflow/5.8.4
module load mysql/8.0.33
```

mkavulich commented:
@climbfuji so I guess that means `module purge` is required for using spack-stack? 

I have omitted ecflow and mysql because we don't use those applications. The new modulefiles appear to be working much better (along with doing a purge first); I pushed the updated Derecho files, and I'll make and test those changes for Hera later. @grantfirl can you try again with the latest files on Derecho (remembering to `module purge` first)?

climbfuji commented:
> @climbfuji so I guess that means `module purge` is required for using spack-stack?
> 
> I have omitted ecflow and mysql because we don't use those applications. The new modulefiles appear to be working much better (along with doing a purge first); I pushed the updated Derecho files, and I'll make and test those changes for Hera later. @grantfirl can you try again with the latest files on Derecho (remembering to `module purge` first)?

Yes - `module purge` comes first.

mkavulich commented:
@climbfuji @grantfirl I am still waiting on help installing LaTeX tools for updating the users guide, but aside from that I think this PR is ready for re-review. I also added modulefiles for Jet and Orion while I was at it since it was simple to add based on the spack-stack instructions Dom sent (I don't have access to any of the other machines).

DomHeinzeller commented:
what tools are you missing?

> On Nov 27, 2023, at 2:10 PM, Michael Kavulich ***@***.***> wrote:
> 
> 
> @climbfuji <https://github.com/climbfuji> @grantfirl <https://github.com/grantfirl> I am still waiting on help installing LaTeX tools for updating the users guide, but aside from that I think this PR is ready for re-review. I also added modulefiles for Jet and Orion while I was at it since it was simple to add based on the spack-stack instructions Dom sent (I don't have access to any of the other machines).
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/NCAR/ccpp-scm/pull/406#issuecomment-1828613847>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AN7FF5DZJBITCXUX6YEWFXTYGT6TLAVCNFSM6AAAAAA7DNMUPCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMRYGYYTGOBUG4>.
> You are receiving this because you were mentioned.
>

climbfuji commented:
I was a professional typesetter for the Journal of Theoretical Astrophysics and other scientific journals for a long time, and we all just used `pdflatex`. Can you try this instead of `latexmk`? Also, the `inc` errors is presumably because a Perl extension is missing.

climbfuji commented:
> I am a bit less experienced as I haven't really used LaTeX since undergrad :)
> 
> The problem appears to exist at the pdflatex level (since it is called by `latexmk`):
> 
> ```
> `I can't find the format file `pdflatex.fmt'!
> ```
> 
> Maybe an issue with the texlive conda package? [conda-forge/texlive-core-feedstock#19](https://github.com/conda-forge/texlive-core-feedstock/issues/19)
> 
> I'm really not inclined to be spending my time debugging this at a deep level since @grantfirl and @dustinswales have been apparently doing this successfully in the past, so I'd like to know how they have done it.

I wonder if we should just defer the documentation build to CI using readthedocs?

grantfirl commented:
@mkavulich We've definitely talked about switching over to sphinx/readthedocs, but haven't gotten around to it. I think that both Dustin and I just used the TeXShop app for Mac and generated the PDF locally. I think it uses pdflatex.

grantfirl commented:
@mkavulich Here is the PDF of the updated docs if you want to include it in this PR:
[main.pdf](https://github.com/NCAR/ccpp-scm/files/13529672/main.pdf)

grantfirl commented:
@mkavulich I'd like to re-test this on Hera and Derecho so that we can maybe get this merged today.

grantfirl commented:
Everything works with Intel/GNU on Hera/Derecho. @mkavulich I'll approve/merge once this is updated to the latest NCAR/main commit.

mkavulich commented:
@grantfirl The branch should now be updated, and I tested one more time on Derecho with Intel. I think it's ready to go 👍

climbfuji commented:
Yay! Welcome to the spack-stack user community :-)
